enhancements/machine-api/machine-pools-in-cluster.md

```diff
@@ -0,0 +1,240 @@
+
+### Goals
+
+List the specific goals of the proposal. How will we know that this has succeeded?
+
+  * Provide customers with an API to automatically spread machines out over the AZs in their account.
+  * Provide OpenShift Hive with a stable API to manage MachineSets spread over AZs day 2.
+
+### Non-Goals
+
+  * No linkup with the MachineDeployment concept for rolling updates of Machines.
+
+## Proposal
+
+### User Stories [optional]
+
+#### Story 1
+
+As an OCP user, I would like to be able to specify min/max machine replicas and
+have those machines automatically spread out amongst all Availability Zones
+that are available in my account.
+
+#### Story 2
+
+As an OpenShift Hive developer, I would like a stable API for spreading
+machines across Availability Zones. This would allow me to sync an API
+in-cluster rather than use one version of vendored code from the installer to
+generate each MachineSet to be reconciled.
+
+### Implementation Details/Notes/Constraints [optional]
+
+Implementation would require a new CRD and controller aligned hopefully with a pre-existing operator in OCP.
+
+The installer MachinePool definition can be seen [here](https://github.com/openshift/installer/blob/master/pkg/types/machinepools.go).
+
+The code for spreading MachineSets across AZs [already exists today](https://github.com/openshift/installer/blob/master/pkg/asset/machines/aws/machinesets.go) and could be removed from the installer, which would then just pass through an instance of the new API.
+
+The [Hive MachinePool API](https://github.com/openshift/hive/blob/master/pkg/apis/hive/v1/machinepool_types.go) also includes support for enabling auto-scaling on the resulting MachineSets, and this functionality should be carried over to the new API even if not initially supported by the Installer.
+
+### Risks and Mitigations
+
+  * Clusters being upgraded would not be making use of this new API. They would continue to have any "orphaned" MachineSets the installer generated for them day 0 but this should work fine.
+
+## Design Details
+
+### Open Questions [optional]
+
+This is where to call out areas of the design that require closure before deciding
+to implement the design.  For instance,
+ > 1. Upstream cluster-api project has begun using the MachinePool naming for an API that is backed by native cloud provider auto-scaling groups. (ASGs, MIGs) We however are already using this term in both the Installer and Hive. Is it acceptable to push this naming in cluster even if it may one day overlap with a cluster-api type in a different apigroup?
```

Comment:
I would prefer we came up with a new name for this as I would expect that at some point we will adopt MachinePools from the cluster-api proejct and I think having two resources in different API groups which do rather different things would be confusing for users

Reply:
Yeah this is really unfortunate. MachinePool is a thing in the installer InstallConfig and in Hive. 

We can work with a new API name though but it's going to be confusing for a long time to come.